base/shared/src/main/scala/scalaz/data/Free.scala

```diff
@@ -0,0 +1,18 @@
+package scalaz
+package data
+
+trait FreeModule {
+  type Free[F[_], A]
+
+  def runFree[F[_], A](f: Free[F, A]): A \/ F[data.Free[F, A]]
+
+  def wrapFree[F[_], A](f: A \/ F[data.Free[F, A]]): Free[F, A]
+}
+
+private[data] object FreeImpl extends FreeModule {
+  type Free[F[_], A] = A \/ F[data.Free[F, A]]
```

Comment:
This representation of `Free` is known to be non-stack-safe (and inefficient).

Reply:
We'll fix this then. Do you recommend a specific paper/project with a more efficient representation?

Reply:
I think scalaz 7 is a good starting point, i.e.

```scala
A | F[A] | (Free[F, X], X => Free[F, A])
```
(where the case `F[A]` represents _embedding_ `F[A]` into `Free[F, A]`, and the other two cases represent the _free versions of operations_ `pure` and `flatMap`, respectively).

In addition to being stack-safe, it also avoids requiring `Functor[F]` for monad operations on `Free[F, A]`.